app/routes/circulars/circulars.server.ts

```diff
@@ -0,0 +1,248 @@
+      ExpressionAttributeValues: searchString
+        ? {
+            ':searchTerm': searchString,
+          }
+        : undefined,
+      FilterExpression: searchString
+        ? 'contains(subject, :searchTerm) OR contains(body, :searchTerm) OR contains(submitter, :searchTerm)'
+        : undefined,
+    })
+    return result
+  }
+
+  /**
+   * Get a specified Circular by Id.
+   *
+   * Throws an HTTP error if:
+   *  - the provided Id does not match any existing circulars
+   *
+   * @param id - the number Id (ex. 12345) of a specific Circular
+   * @returns a CircularModel object
+   */
+  async getCircularById(circularId: number): Promise<CircularModel> {
+    const db = await tables()
+    const result = await db.gcn_circulars.get({
+      circularId: circularId,
+    })
+    if (!result)
+      throw new Response('The requested circular does not exist', {
+        status: 404,
+      })
+    return result
+  }
+}
+
+export class CircularsServer {
+  #sub: string
+  #email: string
+  #name?: string
+  #groups: string[]
+  #affiliation?: string
+
+  private constructor(
+    sub: string,
+    email: string,
+    groups: string[],
+    name?: string,
+    affiliation?: string
+  ) {
+    this.#affiliation = affiliation
+    this.#email = email
+    this.#groups = groups
+    this.#name = name
+    this.#sub = sub
+  }
+
+  static async create(request: Request) {
+    const user = await getUser(request)
+    if (!user)
+      throw new Response('not signed in', {
+        statusText: 'not signed in',
+        status: 403,
+      })
+
+    return new this(
+      user.sub,
+      user.email,
+      user.groups,
+      user.name,
+      user.affiliation
+    )
+  }
+
+  /**
+   * Adds a new entry into the GCN Circulars table
+   *
+   * Throws an HTTP error if:
+   *  - The current user is not in the submitters group
+   *  - Body or Subject are blank
+   * @param body - main content of the Circular
+   * @param subject - the title/subject line of the Circular
+   * @param correctedCircularId - an optional parameter, the Id of an existing circular that this new one updates
+   */
+  async createNewCircular(
+    body: string,
+    subject: string,
+    correctedCircularId?: number
+  ) {
+    if (
+      !this.#groups.find((group) => group === 'gcn.nasa.gov/circular-submitter')
+    )
+      throw new Response('User is not in the submitters group', {
+        status: 403,
+      })
+    if (!body || !subject)
+      throw new Response('Subject and Body cannot be blank', { status: 400 })
+
+    const db = await tables()
+    const result = await db.gcn_circulars.scan({
+      FilterExpression: 'attribute_exists(circularId)',
+      ProjectionExpression: 'circularId',
+    })
+
+    let itemId =
+      (result.Items[0]
+        ? Math.max(...result.Items.map((x) => x['circularId']))
+        : 0) + 1
+    const submitter =
+      this.#name +
+      (this.#affiliation ? ` at ${this.#affiliation} ` : ' ') +
+      this.#email
```

Comment:
This is a very expensive (as in both cycles and $'s) to get the next circular ID. Please follow the optimistic locking example in https://bitesizedserverless.com/bite/reliable-auto-increments-in-dynamodb/, which has the following essential features:

1. A special row in a table that stores the last assigned ID.
2. A query that retrieves that last assigned ID.
3. A transaction that attempts to increment the last assigned ID and put the new record.
4. A loop that retries steps 2 and 3 until the transaction succeeds.